pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,1074 @@
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+		weight_to_charge: Weight,
+	) -> DispatchResult {
+		let fee_currency = schedule.order.get_asset_in();
+		let fee_amount_in_sold_asset = Self::convert_weight_to_fee(weight_to_charge, fee_currency)?;
+
+		Self::unallocate_amount(schedule_id, schedule, fee_amount_in_sold_asset)?;
+
+		T::Currencies::transfer(
+			fee_currency,
+			&schedule.owner,
+			&T::FeeReceiver::get(),
+			fee_amount_in_sold_asset,
+		)?;
+
+		Ok(())
+	}
+
+	fn terminate_schedule(
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+		error: DispatchError,
+	) {
+		Self::try_unreserve_all(schedule_id, schedule);
+
+		Self::remove_schedule_from_storages(&schedule.owner, schedule_id);
+
+		Self::deposit_event(Event::Terminated {
+			id: schedule_id,
+			who: schedule.owner.clone(),
+			error,
+		});
+	}
+
+	fn complete_schedule(schedule_id: ScheduleId, schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>) {
+		Self::try_unreserve_all(schedule_id, schedule);
+
+		Self::remove_schedule_from_storages(&schedule.owner, schedule_id);
+
+		Self::deposit_event(Event::Completed {
+			id: schedule_id,
+			who: schedule.owner.clone(),
+		});
+	}
+
+	fn try_unreserve_all(schedule_id: ScheduleId, schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>) {
+		let sold_currency = schedule.order.get_asset_in();
+
+		let Some(remaining_amount) = RemainingAmounts::<T>::get(schedule_id) else {
+			//Invalid state, we ignore as we terminate the whole DCA anyway
+			return;
+		};
+
+		T::Currencies::unreserve_named(
+			&T::NamedReserveId::get(),
+			sold_currency,
+			&schedule.owner,
+			remaining_amount,
+		);
+	}
+
+	fn weight_to_fee(weight: Weight) -> Balance {
+		// cap the weight to the maximum defined in runtime, otherwise it will be the
+		// `Bounded` maximum of its data type, which is not desired.
+		let capped_weight: Weight = weight.min(T::BlockWeights::get().max_block);
+		<T as pallet::Config>::WeightToFee::weight_to_fee(&capped_weight)
+	}
+
+	fn plan_schedule_for_block(
+		who: T::AccountId,
+		blocknumber: T::BlockNumber,
+		schedule_id: ScheduleId,
+	) -> DispatchResult {
+		let current_block_number = frame_system::Pallet::<T>::current_block_number();
+		ensure!(blocknumber > current_block_number, Error::<T>::BlockNumberIsNotInFuture);
+
+		let next_free_block = Self::find_next_free_block(blocknumber)?;
+
+		if ScheduleIdsPerBlock::<T>::contains_key(next_free_block) {
+			ScheduleIdsPerBlock::<T>::try_mutate_exists(next_free_block, |schedule_ids| -> DispatchResult {
+				let schedule_ids = schedule_ids.as_mut().ok_or(Error::<T>::InvalidState)?;
+
+				schedule_ids
+					.try_push(schedule_id)
+					.map_err(|_| Error::<T>::InvalidState)?;
+				Ok(())
+			})?;
+			return Ok(());
+		} else {
+			let vec_with_first_schedule_id = Self::create_bounded_vec(schedule_id)?;
+			ScheduleIdsPerBlock::<T>::insert(next_free_block, vec_with_first_schedule_id);
+		}
```

Comment:
you can simplify this whole block as follows:

```suggestion
		ScheduleIdsPerBlock::<T>::try_mutate(next_free_block, |schedule_ids| -> DispatchResult {
			schedule_ids
				.try_push(schedule_id)
				.map_err(|_| Error::<T>::InvalidState)?;
			Ok(())
		})?;
```

benefits:
1. before you did not deposit event in the case where the schedule id is already in the list
2. you get rid of the invalid states
3. you can remove the explicit create_bounded_vec shenanigans.